File: 132.cpp
struct Solution {
    bool ispal(int i, int j, vector<vector<char>> &rec, string &s){
        if (i > j){
            return true;
        }
        if (rec[i][j] == 2){
            if (s[i] != s[j]){
                rec[i][j] = 0;
            } else {
                rec[i][j] = this->ispal(i + 1, j - 1, rec, s);
            }
        }
        return rec[i][j] == 1;
    }
    int minCut(string &s) {
        int n = s.size();
        vector<vector<char>> rec(n, vector<char>(n, 2));
        for (int i = 0; i < n; ++i){
            rec[i][i] = 1;
        }
        if (this->ispal(0, n - 1, rec, s)){
            return 0;
        }
        vector<int> memo(n, n);
        memo[0] = 0;
        for (int i = 1; i < n; ++i){
            if (this->ispal(0, i, rec, s)){
                memo[i] = 0;
                continue;
            }
            for (int j = 0; j < i; ++j){
                if (this->ispal(j + 1, i, rec, s)){
                    memo[i] = min(memo[i], memo[j] + 1);
                }
            }
        }
        return memo.back();
    }
};

File: sol/132.py
from functools import cache

class Solution:
    def minCut(self, s: str) -> int:
        @cache
        def ispal(i, j):
            if i >= j:
                return True
            return s[i] == s[j] and ispal(i + 1, j - 1)
        n = len(s)
        if ispal(0, n - 1):
            return 0
        memo = [n - 1] * n
        memo[0] = 0
        for i in range(1, n):
            if ispal(0, i):
                memo[i] = 0
                continue
            for j in range(i):
                if ispal(j + 1, i):
                    memo[i] = min(memo[i], memo[j] + 1)
        return memo[n - 1]
